### ISSUE TITLE
`pool.terminate()` crashing Electron app

### ISSUE THREAD

**Issue by sftsk** [andywer/threads.js#247]
Hello,

thanks for this great library. I really like the syntax and everything seems to be doing exactly what I want.
Unfortunately I have an issue with the `pool.terminate()` command. If I use it, it pretty much every time crashes the Electron Renderer process. Any idea what might be the issue? I don't get any js/ts error. In the crash reports I can see the following error: `EXC_BAD_ACCESS / EXC_I386_GPFLT`.

I am on Electron `8.3.0` and using it with Electron Forge `6.0.0-beta.51` and the webpack/typescript/react template.

This is the code and I made sure it's only run once:

```
const loadCsvItemModels = async () => {
    const workerPool = Pool(
      () => spawn(new Worker("../../workers/parseStl")));

    filteredItems.forEach((item) => {
      // load stl from csv
      if (item.stl !== undefined && item._sam.model === undefined) {
        if (fs.existsSync(item.stl)) {
          setStoreState({
            type: "SET_ITEM_MODEL_DATA",
            sku: item.sku,
            modelData: { isComputing: true },
          });
          workerPool.queue(async (parseStl: any) => {
            const modelData = await parseStl.getModelData({
              stlPath: item.stl
            });
            storeResults(item.sku, modelData);
          });
        }
      }
    });

    // await workerPool.completed();
    workerPool.terminate();
  };
```

If I leave out `terminate()` everything works as expected. Is there any downside/ memory leak possibility if I just leave it out? It seems like the pool is terminating anyway if there is nothing added to the queue after the specified timeout.

Thank you for your help!

**Comment by andywer** [comment]
Hey @sftsk!

I suppose it's due to the out-commented `await workerPool.completed()` line. You kill the pool while it's still doing things. Or does it also crash when you await the pool completion first?

PS: Try opening the developer tools of the window and run this in the console: `localStorage.setItem("debug", "threads*")`. This should enable debug logging and might reveal some useful information.

**Comment by sftsk** [comment]
Hello @andywer , thanks for the swift reply. It also crashes when leaving in the `await workerPool.completed();`, I just tried removing it to see if anything improves.

I have activated the debug but can't spot anything off.

This is the log:

```
threads:master:spawn Initializing new thread +0ms
instrument.js:102 threads:master:spawn Initializing new thread +5ms
6instrument.js:102 threads:master:spawn Initializing new thread +1ms
instrument.js:102 threads:master:spawn Initializing new thread +2ms
3instrument.js:102 threads:master:spawn Initializing new thread +1ms
instrument.js:102 threads:pool:1 Queueing task #1... +0ms
instrument.js:102 threads:pool:1 Attempt de-queueing a task in order to run it... +0ms
instrument.js:102 threads:pool:1 Queueing task #2... +3ms
instrument.js:102 threads:pool:1 Attempt de-queueing a task in order to run it... +0ms
instrument.js:102 threads:pool:1 Queueing task #3... +2ms
instrument.js:102 threads:pool:1 Attempt de-queueing a task in order to run it... +0ms
instrument.js:102 threads:pool:1 Queueing task #4... +2ms
instrument.js:102 threads:pool:1 Attempt de-queueing a task in order to run it... +0ms
instrument.js:102 threads:pool:1 Queueing task #5... +2ms
instrument.js:102 threads:pool:1 Attempt de-queueing a task in order to run it... +0ms
instrument.js:102 threads:pool:1 Queueing task #6... +2ms
instrument.js:102 threads:pool:1 Attempt de-queueing a task in order to run it... +0ms
instrument.js:102 threads:pool:1 Queueing task #7... +2ms
instrument.js:102 threads:pool:1 Attempt de-queueing a task in order to run it... +0ms
instrument.js:102 threads:pool:1 Queueing task #8... +2ms
instrument.js:102 threads:pool:1 Attempt de-queueing a task in order to run it... +0ms
instrument.js:102 threads:pool:1 Queueing task #9... +1ms
instrument.js:102 threads:pool:1 Attempt de-queueing a task in order to run it... +1ms
instrument.js:102 threads:pool:1 Running task #1 on worker #1... +94ms
instrument.js:102 threads:pool:1 Running task #2 on worker #2... +1ms
instrument.js:102 threads:pool:1 Running task #3 on worker #3... +0ms
instrument.js:102 threads:pool:1 Running task #4 on worker #4... +0ms
instrument.js:102 threads:pool:1 Running task #5 on worker #5... +0ms
instrument.js:102 threads:pool:1 Running task #6 on worker #6... +1ms
instrument.js:102 threads:pool:1 Running task #7 on worker #7... +0ms
instrument.js:102 threads:pool:1 Running task #8 on worker #8... +0ms
instrument.js:102 threads:pool:1 Running task #9 on worker #9... +1ms
instrument.js:102 threads:master:messages Message from worker before finishing initialization: +0ms {type: "init", exposed: {…}}
instrument.js:102 queuing: Item 9
instrument.js:102 threads:master:messages Sending command to run function to worker: +0ms {type: "run", uid: 1, method: "getModelData", args: Array(1)}
instrument.js:102 threads:master:messages Message from worker before finishing initialization: +2ms {type: "init", exposed: {…}}
instrument.js:102 queuing: Item 8
instrument.js:102 threads:master:messages Sending command to run function to worker: +2ms {type: "run", uid: 2, method: "getModelData", args: Array(1)}
instrument.js:102 threads:master:messages Message from worker before finishing initialization: +1ms {type: "init", exposed: {…}}
instrument.js:102 queuing: Item 7
instrument.js:102 threads:master:messages Sending command to run function to worker: +1ms {type: "run", uid: 3, method: "getModelData", args: Array(1)}
instrument.js:102 threads:master:messages Message from worker before finishing initialization: +1ms {type: "init", exposed: {…}}
instrument.js:102 queuing: Item 6
instrument.js:102 threads:master:messages Sending command to run function to worker: +1ms {type: "run", uid: 4, method: "getModelData", args: Array(1)}
instrument.js:102 threads:master:messages Message from worker: +144ms {type: "running", uid: 4, resultType: "promise"}
instrument.js:102 threads:master:messages Message from worker before finishing initialization: +164ms {type: "init", exposed: {…}}
instrument.js:102 queuing: Item 2
instrument.js:102 threads:master:messages Sending command to run function to worker: +20ms {type: "run", uid: 5, method: "getModelData", args: Array(1)}
instrument.js:102 threads:master:messages Message from worker before finishing initialization: +9ms {type: "init", exposed: {…}}
instrument.js:102 queuing: Item 4
instrument.js:102 threads:master:messages Sending command to run function to worker: +9ms {type: "run", uid: 6, method: "getModelData", args: Array(1)}
instrument.js:102 threads:master:messages Message from worker: +22ms {type: "running", uid: 2, resultType: "promise"}
instrument.js:102 threads:master:messages Message from worker: +2ms {type: "result", uid: 4, complete: true, payload: {…}}
instrument.js:102 Item 6 {path: "/Users/sftsk/Documents/industrial/kammio_fixed2010.stl", boundingBox: Array(3), volume: 143.25}
instrument.js:102 threads:pool:1 Task #4 completed successfully +608ms
instrument.js:102 threads:pool:1 Attempt de-queueing a task in order to run it... +0ms
instrument.js:102 threads:pool:1 Task queue is empty +0ms
instrument.js:102 threads:master:messages Message from worker before finishing initialization: +28ms {type: "init", exposed: {…}}
instrument.js:102 queuing: Item 1
instrument.js:102 threads:master:messages Sending command to run function to worker: +4ms {type: "run", uid: 7, method: "getModelData", args: Array(1)}
instrument.js:102 threads:master:messages Message from worker before finishing initialization: +7ms {type: "init", exposed: {…}}
instrument.js:102 queuing: Item 5
instrument.js:102 threads:master:messages Sending command to run function to worker: +6ms {type: "run", uid: 8, method: "getModelData", args: Array(1)}
instrument.js:102 threads:master:messages Message from worker: +8ms {type: "running", uid: 8, resultType: "promise"}
instrument.js:102 threads:master:messages Message from worker: +15ms {type: "result", uid: 8, complete: true, payload: {…}}
instrument.js:102 Item 5 {path: "/Users/sftsk/Documents/industrial/gearren.stl", boundingBox: Array(3), volume: 4.25}
instrument.js:102 threads:pool:1 Task #5 completed successfully +32ms
instrument.js:102 threads:pool:1 Attempt de-queueing a task in order to run it... +1ms
instrument.js:102 threads:pool:1 Task queue is empty +0ms
instrument.js:102 threads:master:messages Message from worker: +38ms {type: "result", uid: 2, complete: true, payload: {…}}
instrument.js:102 Item 8 {path: "/Users/sftsk/Documents/industrial/Steering Mount.stl", boundingBox: Array(3), volume: 43.07}
instrument.js:102 threads:pool:1 Task #2 completed successfully +37ms
instrument.js:102 threads:pool:1 Attempt de-queueing a task in order to run it... +0ms
instrument.js:102 threads:pool:1 Task queue is empty +1ms
instrument.js:102 threads:master:messages Message from worker before finishing initialization: +108ms {type: "init", exposed: {…}}
instrument.js:102 queuing: Item 3
instrument.js:102 threads:master:messages Sending command to run function to worker: +48ms {type: "run", uid: 9, method: "getModelData", args: Array(1)}
instrument.js:102 threads:master:messages Message from worker: +31ms {type: "running", uid: 9, resultType: "promise"}
instrument.js:102 threads:master:messages Message from worker: +10ms {type: "running", uid: 6, resultType: "promise"}
instrument.js:102 threads:master:messages Message from worker: +29ms {type: "result", uid: 9, complete: true, payload: {…}}
instrument.js:102 Item 3 {path: "/Users/sftsk/Documents/industrial/Cylinder robic.stl", boundingBox: Array(3), volume: 10.67}
instrument.js:102 threads:pool:1 Task #7 completed successfully +117ms
instrument.js:102 threads:pool:1 Attempt de-queueing a task in order to run it... +0ms
instrument.js:102 threads:pool:1 Task queue is empty +0ms
instrument.js:102 threads:master:messages Message from worker before finishing initialization: +109ms {type: "init", exposed: {…}}
instrument.js:102 threads:master:messages Message from worker before finishing initialization: +22ms {type: "init", exposed: {…}}
instrument.js:102 threads:master:messages Message from worker: +119ms {type: "running", uid: 1, resultType: "promise"}
instrument.js:102 threads:master:messages Message from worker: +9ms {type: "running", uid: 7, resultType: "promise"}
instrument.js:102 threads:master:messages Message from worker before finishing initialization: +72ms {type: "init", exposed: {…}}
instrument.js:102 threads:master:messages Message from worker: +23ms {type: "result", uid: 1, complete: true, payload: {…}}
instrument.js:102 Item 9 {path: "/Users/sftsk/Documents/industrial/Turbine Blade.stl", boundingBox: Array(3), volume: 1455.23}
instrument.js:102 threads:pool:1 Task #1 completed successfully +151ms
instrument.js:102 threads:pool:1 Attempt de-queueing a task in order to run it... +0ms
instrument.js:102 threads:pool:1 Task queue is empty +0ms
instrument.js:102 threads:master:messages Message from worker: +3ms {type: "result", uid: 7, complete: true, payload: {…}}
instrument.js:102 Item 1 {path: "/Users/sftsk/Documents/industrial/blade.stl", boundingBox: Array(3), volume: 968.75}
instrument.js:102 threads:pool:1 Task #9 completed successfully +2ms
instrument.js:102 threads:pool:1 Attempt de-queueing a task in order to run it... +0ms
instrument.js:102 threads:pool:1 Task queue is empty +1ms
instrument.js:102 threads:master:messages Message from worker: +24ms {type: "result", uid: 6, complete: true, payload: {…}}
instrument.js:102 Item 4 {path: "/Users/sftsk/Documents/industrial/cylindrical_gyroid_15mm_thick_walls.stl", boundingBox: Array(3), volume: 5.56}
instrument.js:102 threads:pool:1 Task #6 completed successfully +24ms
instrument.js:102 threads:pool:1 Attempt de-queueing a task in order to run it... +0ms
instrument.js:102 threads:pool:1 Task queue is empty +0ms
instrument.js:102 threads:master:messages Message from worker: +139ms {type: "running", uid: 5, resultType: "promise"}
instrument.js:102 threads:master:messages Message from worker: +31ms {type: "result", uid: 5, complete: true, payload: {…}}
instrument.js:102 Item 2 {path: "/Users/sftsk/Documents/industrial/crank shaft.stl", boundingBox: Array(3), volume: 3073.34}
instrument.js:102 threads:pool:1 Task #8 completed successfully +169ms
instrument.js:102 threads:pool:1 Attempt de-queueing a task in order to run it... +1ms
instrument.js:102 threads:pool:1 Task queue is empty +0ms
instrument.js:102 threads:master:messages Message from worker: +1s {type: "running", uid: 3, resultType: "promise"}
instrument.js:102 threads:master:messages Message from worker: +91ms {type: "result", uid: 3, complete: true, payload: {…}}
instrument.js:102 Item 7 {path: "/Users/sftsk/Documents/industrial/manifold.stl", boundingBox: Array(3), volume: 817.82}
instrument.js:102 threads:pool:1 Task #3 completed successfully +1s
instrument.js:102 threads:pool:1 Attempt de-queueing a task in order to run it... +1ms
instrument.js:102 threads:pool:1 Task queue is empty +0ms
3instrument.js:102 threads:master:thread-utils Terminating worker +0ms
instrument.js:102 threads:master:thread-utils Terminating worker +1ms
instrument.js:102 threads:master:thread-utils Terminating worker +0ms
instrument.js:102 threads:master:thread-utils Terminating worker +1ms
instrument.js:102 threads:master:thread-utils Terminating worker +0ms
```

I have simplified the code a tiny bit for the test and not storing the results right now:

```
const loadCsvItemModels = async () => {
    const workerPool = Pool(() => spawn(new Worker("../../workers/parseStl")));

    filteredItems.forEach((item) => {
      // load stl from csv
      if (item.stl !== undefined && item._sam.model === undefined) {
        if (fs.existsSync(item.stl)) {  
          workerPool.queue(async (parseStl: any) => {
            console.log("queueing: " + item.sku);
            const modelData = await parseStl.getModelData({
              stlPath: item.stl,
            });
            console.log(item.sku, modelData);
            // storeResults(item.sku, modelData);
          });
        }
      }
    });
    await workerPool.completed();
    workerPool.terminate();
  };
```

Does this debug give you any insights?

**Comment by andywer** [comment]
Yeah, debug log looks pretty unsuspicious… I looked up `EXC_BAD_ACCESS` and it might hint us towards the issue: The error indicates that [you are sending a message to an object that has already been released](https://code.tutsplus.com/tutorials/what-is-exc_bad_access-and-how-to-debug-it--cms-24544).

So I guess we are still posting messages to workers after we terminated them. Can you maybe just quickly monkey-patch the library code in your `node_modules/` and add a `console.trace("postMessage():", args)` [here](https://github.com/andywer/threads.js/blob/master/src/master/invocation-proxy.ts#L120)?

Would be interesting to know if
    a) this is really the cause and
    b) if so, why there is another invocation after `workerPool.terminate()`

**Comment by sftsk** [comment]
I tried adding this trace. It doesn't fire after the crash.
If I add a console log in between the `await workerPool.completed()` and the `workerPool.terminate()` the console log still fires before the crash.

It does feel like I am trying to fire `workerPool.terminate()` on an instance that doesn't exit anymore :-/ could it be terminate() internally tries calling something that might already have vanished from memory?

**Comment by sftsk** [comment]
Hello,

I have tried digging deeper but can't dig any deeper as I don't fully understand the code.
On this line: https://github.com/andywer/threads.js/blob/73c9daa35809d8071dda48bc9a61910ca6ec336c/src/master/spawn.ts#L114 the error seems to be happening. If I take it out, the app won't crash (and also the worker won't terminate I suppose..). Any pointer on how I can debug further from here?

Thanks for your help, would really like to find a solution to this.

**Comment by andywer** [comment]
Thanks for helping out here, @sftsk!

Can you try to monkey-patch the following quick-fix into your `master/spawn.js` file?

```diff
        debugThreadUtils("Terminating worker");
        // Newer versions of worker_threads workers return a promise
+       if (!worker._is_terminated) {
+       worker._is_terminated = true
        yield worker.terminate();
+       }
        resolver();
```

I suppose the termination method is called more than once – which should be fine (the second call should just do nothing), but maybe there is some weird hiccup in the underlying implementation.

**Comment by sftsk** [comment]
Hello, unfortunately this has not helped :(

**Comment by andywer** [comment]
@sftsk `thread[$terminate]` is set to this function while spawning a thread: https://github.com/andywer/threads.js/blob/73c9daa35809d8071dda48bc9a61910ca6ec336c/src/master/spawn.ts#L111 (a.k.a. the function we already looked at)

I think you will have to create a repository with a minimal setup to recreate the issue. Otherwise it will be very hard to help you. Also, since the issue hasn't yet been confirmed by someone else, it could also be an issue with your app's code (even though I think it's unlikely).

**Comment by sftsk** [comment]
I will try to set up something minimal and see if it helps to get to the bottom of the issue.
Thanks for the help already!

One last question: if I don’t call terminate, will I eventually have an issue with open listeners or blocked memory or something?

Because as said, not calling the terminate function everything seems to be working just fine..

**Comment by andywer** [comment]
> if I don’t call terminate, will I eventually have an issue with open listeners or blocked memory or something?

Well, the workers will needlessly consume a bunch of memory. Not sure if it’s significant, though – might do for the time being 😉

**Comment by sftsk** [comment]
Hello, I have set up a tiny project: https://github.com/sftsk/electron-react-ts-boilerplate-threads  based of this [template](https://github.com/euZebe/electron-react-ts-boilerplate).

I have added `threads` and used a pool. The worker uses the demo `add` function you have in the getting started notes. This tiny demo consistently crashes the electron renderer process for me.

I have noticed if I turn off `nodeIntegrationInWorker` in the electron `webPreferences` the app won't crash. Seems like the crash is somehow related to that :-/ Unfortunately I need that in my actual app. Would be great if you can figure out anything with this input.

**Comment by andywer** [comment]
Cool! I will look into it as soon as I find some time.

Just as a general note: Of course it shouldn't crash the process either way, but think twice if you really need the `nodeIntegrationInWorker` as it's a really bad decision from an app security point of view 😉

**Comment by andywer** [comment]
Also, I am just realizing:

Using `threads` in the renderer process with `nodeIntegrationInWorker` enabled is quite a tricky edge case, anyway. Should `threads` use the web view's web worker implementation or node.js' `worker_threads` module? 🤔

**Comment by sftsk** [comment]
Hello again, I have decided to refactor my code to not need the node integration inside the workers. Now everything is stable and the issue is closed from my side.

Thanks for the help and I hope this might other people to the right direction as well :)

**Comment by andywer** [comment]
Great! 🙂

**Comment by Aaronaught** [comment]
I ran into a very similar issue with a simpler app, relatively speaking, and I suspect it's probably the same bug.

- I'm not using pooling, just `await Thread.terminate`
- The worker and main thread use typeorm for DB access and winston for logging (not trying to share instances or anything weird like that)
- The main thread is an Express HTTP handler
- I noticed that the HTTP handler often gets its result while the worker is still spewing log output in the console - I guess Winston is buffering the logs and writing them late, and maybe this is somewhat related. Usually the process would crash the first time running the worker if this happens.
- Sometimes the worker would complete, either by accident or because I introduced some artificial waits. In these cases, it would always, without exception, crash on the second attempt, *before* ever actually running any worker code. With logging for threads.js turned on, I can see that the very last thing that happens before the crash is "initializing new thread".
- Most of the time I get no useful output from the crash, but once in a while I would see a stack trace to `napi_create_reference`.
- It's not _specifically_ Winston that's the issue, because I even if I take out all of the logging code and dependencies, it still crashes - without the worker logging it will generally get through the first attempt but crash on the second, again before ever actually running any code in the second thread.

At first I thought it was related to typeorm because I've seen the SQLite driver just crash the app outright (no stack trace, no nothing) in cases where e.g. the file is inaccessible or something is misconfigured. So, similar behavior. But I've confirmed that it doesn't even make it to the point where it tries to open any database, at least in the second-run crashes. It must be a more general promise/async/await type of thing.

All I can surmise is that node and/or threads.js becomes very unstable if a "thread" is "terminated" while it is doing any work, and that even if I make sure to eliminate all the visible work and clean up all explicit resources, there is still some implicit stuff going on that I can't control. And the way this instability manifests, it may either crash when terminating the first worker, or if not, it will crash when starting another instance of the worker. If I don't call `Thread.terminate`, everything is fine, so I don't think it's a matter of orphaned/leaked resources, otherwise it would crash even without the `terminate`.

I can even run two workers at the same time, starting from two separate HTTP requests, which rules out quite a few potential explanations. Whatever the problem is, it's definitely somewhere in `terminate`.
